File: OOP Start/PhonebookApp/Core/Infrastructure/Phonebook.java
package PhonebookApp.Core.Infrastructure;

import java.util.ArrayList;
import java.util.List;

import PhonebookApp.Core.Models.Contact;

public class Phonebook {
    private List<Contact> contacts;
    
    public Phonebook() {
        contacts = new ArrayList<Contact>();
    }

    // create
    public boolean add(Contact contact) {
        if (!contacts.contains(contact)) {
            contacts.add(contact);
            return true;
        }

        return false;
    }

    // join current contacts with other phonebook contacts
    public boolean join(Phonebook phonebook) {
        if (phonebook != null) {
            return addAll(phonebook.getContacts());
        }

        return false;
    }

    // join current contacts with other phonebook contacts
    public boolean addAll(List<Contact> contacts) {
        if (contacts != null) {
            for (Contact contact : contacts) {
                this.contacts.add(contact);
            }

            return true;
        }

        return false;
    }

    // read
    public Contact getContact(int index) {
        return contains(index) ? contacts.get(index) : null;
    }

    public int getContactIndex(Contact contact) {
        return contacts.indexOf(contact);
    }

    // update
    // ???...

    // delete
    public boolean remove(Contact contact) {
        if (getContactIndex(contact) != -1) {
            contacts.remove(contact);
            return true;
        }

        return false;
    }


    private boolean contains(int index) {
        return contacts != null && contacts.size() > index;
    }

    public List<Contact> getContacts() {
        return contacts;
    }

    public List<Contact> search(String[] query) {
        List<Contact> result = new ArrayList<Contact>();
        if (query != null && query.length > 0) {
            for (Contact contact : contacts) {
                if (contact.toString().toLowerCase().contains(query[0].toLowerCase())) {
                    result.add(contact);
                }
            }
        }

        return result;
    }

    public int count() {
        return contacts.size();
    }
}

File: OOP Start/OOP. Lesson 7/Core/Infrastructure/Phonebook.py
from ..Models.Contact import Contact

class Phonebook:
    def __init__(self):
        self.contacts = []

    # create
    def add(self, contact: Contact) -> bool:
        if contact not in self.contacts:
            self.contacts.append(contact)
            return True
        return False

    # join current contacts with other phonebook contacts
    def join(self, phonebook: 'Phonebook') -> bool:
        if phonebook is not None:
            return self.add_all(phonebook.get_contacts())
        return False

    # join current contacts with other phonebook contacts
    def add_all(self, contacts) -> bool:
        if contacts is not None:
            for contact in contacts:
                self.contacts.append(contact)
            return True
        return False

    # read
    def get_contact(self, index: int) -> Contact:
        return self.contacts[index] if self.contains(index) else None

    def get_contact_index(self, contact: Contact) -> int:
        return self.contacts.index(contact) if contact in self.contacts else -1

    # delete
    def remove(self, contact: Contact) -> bool:
        if self.get_contact_index(contact) != -1:
            self.contacts.remove(contact)
            return True
        return False

    def contains(self, index: int) -> bool:
        return self.contacts is not None and len(self.contacts) > index

    def get_contacts(self):
        return self.contacts

    def search(self, query):
        result = []
        if query is not None and len(query) > 0:
            for contact in self.contacts:
                if query[0].lower() in str(contact).lower():
                    result.append(contact)
        return result

    def count(self) -> int:
        return len(self.contacts)
